File: ocaml/bin/p0025.ml
(* Project Euler: Problem 25 *)

open Core

let calc_by_zarith () =
  let ulimit = Z.pow (Z.of_int 10) 999 in    (* the minimum number which contains 1000 digits is 10**999. *)
  let rec loop f1 f2 idx =
    if Z.geq f2 ulimit then
      idx
    else
      loop f2 Z.(f2 + f1) (succ idx)
  in
  loop Z.one Z.one 2
  
let calc_by_array () =
  let fib = Array.make_matrix ~dimx:3 ~dimy:(1000 + 1) 0 in    (* '+ 1' is a spare area that will never be accessed. *)
  let i = ref 0 in
  let cnt = ref 2 in
  let carry = ref 0 in
  fib.(0).(0) <- 1; fib.(1).(0) <- 1; fib.(2).(0) <- 1;
  while fib.(!i).(999) = 0 do
    i := (!i + 1) mod 3;
    cnt := !cnt + 1;
    for j = 0 to 999 do
      fib.(!i).(j) <- fib.((!i + 1) mod 3).(j) + fib.((!i + 2) mod 3).(j) + !carry;
      carry := fib.(!i).(j) / 10;
      fib.(!i).(j) <- fib.(!i).(j) mod 10
    done
  done;
  !cnt

let exec () =
  sprintf "%d (w/ Zarith module)\n%d (w/o Zarith module)" (calc_by_zarith ()) (calc_by_array ())

let () = Euler.Task.run exec

File: julia/Euler/src/bin/p0025.jl

# project euler: problem 25

#=
  f₁ = 1, f₂ = 1, f₃ = 2, f₄ = 3, f₅ = 5, f₆ = 8, f₇ = 13, f₈ = 21, f₉ = 34, ...
=#

module Prob0025

function solve_0025(digits = 1_000)
    boundary = big(10) ^ (digits - 1)
    fₙ = big(1); fₙ₋₁ = big(1)
    idx = 2
    while fₙ < boundary
        idx += 1
        (fₙ, fₙ₋₁) = (fₙ + fₙ₋₁, fₙ)
    end
    idx
end

end #module

using .Prob0025: solve_0025
export solve_0025
